### PR TITLE
add rotation of drawing in 90 degree increments

### PR DESCRIPTION
**Pull request by stombeur**
I wanted to be able to plot a prortrait A3 without having to draw it sideways first, and thought that rotating the drawing on the page by 90 degrees would be the best option.

### CODE REVIEW

File: src/massager.ts
Comment (nornagon) [inline src/massager.ts:29]: Hm, it's surprising to me that this needs to be done after scaling. Wouldn't this make the drawing run off the edge of the page?

Further, if this was done prior to scaling, it wouldn't matter what the center of the rotation was.

**Comment by stombeur** [comment]
I tried that, but before scaling it always went off the page unless fit to page was true. I’ll put some more time in it tomorrow and update here.

Stephane Tombeur
________________________________
Van: Jeremy Rose <notifications@github.com>
Verzonden: Saturday, March 6, 2021 8:31:27 AM
Aan: nornagon/saxi <saxi@noreply.github.com>
CC: Stephane Tombeur <stephane@tombeur.be>; Author <author@noreply.github.com>
Onderwerp: Re: [nornagon/saxi] add rotation of drawing in 90 degree increments (#76)


@nornagon commented on this pull request.

________________________________

In src/massager.ts<https://github.com/nornagon/saxi/pull/76#discussion_r588847432>:

> @@ -23,6 +24,18 @@ export function replan(inPaths: Vec2[][], planOptions: PlanOptions): Plan {
     }
   }

+  // Rotate drawing around center of paper to handle plotting portrait drawings
+  // along y-axis of plotter
+  // This must be done after scaling


Hm, it's surprising to me that this needs to be done after scaling. Wouldn't this make the drawing run off the edge of the page?

Further, if this was done prior to scaling, it wouldn't matter what the center of the rotation was.

**Comment by stombeur** [comment]
I think I made it better. Put it back before scaling like you said.

During scaling (when fitToPage = false), you do
`paths = paths.map(ps => ps.map(p => vmul(p, mmPerSvgUnit)))`
and that made me realize that the center of rotation should be expressed in SvgUnits, so `vmul(center, 1/mmPerSvgUnit)`

As for your other remark, I think the center will probably matter if fitToPage=false, because in that case if the center is far off the center of the drawing it will move off the page, so I left that in. Another option would be to calculate the center of the drawing with min/max of its coörds, but that would change the layout on page if it was intentional.

Anyway, no worries if this does not fit the philosophy of your excellent program, I can run from my fork if I need it. In that case, feel free to just close this :-). Thanks for looking at this, btw!

**Comment by jvolker** [comment]
This is great. Thanks for writing this PR. 

Two small changes that I think would improve UX slightly:
- In my opinion, the UI element should move out of "more" up to the "paper" section though as it's a layout option like all other options in "paper".
- It would be handy if it was possible to continuously rotate, wrapping between 270° and 0°. The next value beyond 270° would be 0°. The next value below 0° would be 270°.

**Comment by stombeur** [comment]
I moved it into the paper section, below width/height settings as that also swaps portrait/landscape and seemed appropriate to put those together.
I found a stupid way to wrap around the values, but it works. A bit surprised to find that this is not something standard.
Note that nothing prevents the user to fill in arbitrary values between 0 and 270, but that seems ok to me.

**Review by jvolker — COMMENTED** [review]
Thanks for the changes. That's great.

> I found a stupid way to wrap around the values, but it works. A bit surprised to find that this is not something standard.
Note that nothing prevents the user to fill in arbitrary values between 0 and 270, but that seems ok to me.

Since users might use other rotation value other than multiples of 90° may I suggest to rather do something along those lines:

```
value = value % 360;
if (value < 0) value = value + 360;
```

**Comment by stombeur** [comment]
> Since users might use other rotation value other than multiples of 90° may I suggest to rather do something along those lines:
> 
> ```
> value = value % 360;
> if (value < 0) value = value + 360;
> ```

wouldn't that be annoying for the user? Users will be mostly technical I presume, otherwise they would be using something else to plot, so I would not want to hold their hands and wave a finger when they do something "wrong". 

When I started this feature, I wanted something along the lines of the autorotate feature in the inkscape plugin by emsl. Maybe I should rewrite this to be more like that? Just a checkbox -> autorotate, with the intention of aligning the long axis of the page with the long axis of the plotter. The drawing would not be rotated on-screen, but only when sent to the plotter. Would that be better?

**Comment by jvolker** [comment]
> wouldn't that be annoying for the user? 

What exactly could be annoying?

> Would that be better?

I'm not sure if I would use auto-rotate. As long as it would still be possible to rotate manually it might still be a good addition for some people. I would prefer rotating manually though. Plotting is never a quick process for me and I want more control.

**Comment by stombeur** [comment]
Hi Jeremias,

> What exactly could be annoying?
Typing something and then having it changed by the software without a visual cue or explanation does not feel right to me. Since it does not break anything, I would be fine with allowing 89 degrees to rotate instead of 90.

Anyway, I'm going to stop changing this PR now. I'll let @nornagon decide what to do with this. I'll run from my own fork for now, so no rush. And if you (@nornagon) want any last changes before merging, I'm happy to oblige.

Thanks for this library btw Jeremy, if you have a kofi or something like it, feel free to point me to it so I can buy you a coffee or a beer.

**Comment by jvolker** [comment]
> Typing something and then having it changed by the software without a visual cue or explanation does not feel right to me. 

I agree. Maybe this was a misunderstanding. The code I proposed was supposed to run only when pressing the up/down buttons, not when typing something in.

> Anyway, I'm going to stop changing this PR now. 

Fair enough.

> I'll let @nornagon decide what to do with this. I'll run from my own fork for now, so no rush. And if you (@nornagon) want any last changes before merging, I'm happy to oblige.

I hope I didn't intrude here. I just wanted to give some useful feedback. Happy to step back now.

Again, thanks for your work and changes @stombeur.

**Review by nornagon — APPROVED** [review]

**Comment by nornagon** [comment]
I like this. Thanks @stombeur :)

**Pull request merged**
